src/library_wasmfs_js_file.js

```diff
@@ -0,0 +1,53 @@
+mergeInto(LibraryManager.library, {
+  // JSFile backend: Store a file's data in JS. We map File objects in C++ to
+  // entries here that contain typed arrays.
+  $wasmFS$JSMemoryFiles: {},
+
+  _wasmfs_backend_add_js_file__deps: [
+    '$wasmFS$backends',
+    '$wasmFS$JSMemoryFiles',
+  ],
+  _wasmfs_backend_add_js_file: function(backend) {
+    wasmFS$backends[backend] = {
+      constructor: function(file) {
+        // Do nothing: we allocate the typed array lazily, see write()
+      },
+      destructor: function(file) {
+        // Release the memory, as it now has no references to it any more.
+        wasmFS$JSMemoryFiles[file] = undefined;
+      },
+      write: function(file, buffer, length, offset) {
+        try {
+          if (!wasmFS$JSMemoryFiles[file]) {
+            // Initialize typed array on first write operation.
+            wasmFS$JSMemoryFiles[file] = new Uint8Array(offset + length);
+          }
+
+          if (offset + length > wasmFS$JSMemoryFiles[file].length) {
+            // Resize the typed array if the length of the write buffer exceeds its capacity.
+            var oldContents = wasmFS$JSMemoryFiles[file];
+            var newContents = new Uint8Array(offset + length);
+            newContents.set(oldContents);
+            wasmFS$JSMemoryFiles[file] = newContents;
+          }
+
+          wasmFS$JSMemoryFiles[file].set(HEAPU8.subarray(buffer, buffer + length), offset);
+          return 0;
+        } catch (err) {
+          return {{{ cDefine('EIO') }}};
```

Comment:
Are there any situations where we would expect this code to throw? If not, it would be better to make this an assertion failure rather than propagating the error to user code.

Reply:
This code is only moved without changes from the existing code - it's a pure refactoring. I don't know the answer to your question, sorry.

Reply:
(But I'd guess an allocation error is the main thing here?)